Bottom: eb30e5268c11250ea3208b76647e908b8b4344fd
Top:    9b273587dfc90ac6f03d40c050cd21f621295da7
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2016-12-04 19:16:19 +0000

Fixup line rubber band code to avoid going nuts when rat-lines are present.


---

diff --git a/src/crosshair.c b/src/crosshair.c
index 274ce69..b2a596f 100644
--- a/src/crosshair.c
+++ b/src/crosshair.c
@@ -438,10 +438,11 @@ XORDrawMoveOrCopyObject (hidGC gc)
         i = Crosshair.AttachedObject.RubberbandN;
         ptr = Crosshair.AttachedObject.Rubberband;
 
-        while (i)
+        for (; i; ptr++, i--)
           {
             if (!TEST_FLAG (VIAFLAG, ptr->Line) &&
-                TEST_FLAG (RUBBERENDFLAG, ptr->Line))
+                TEST_FLAG (RUBBERENDFLAG, ptr->Line) &&
+                ptr->Layer != NULL)
               {
                 double multiplier;
 
@@ -471,9 +472,6 @@ XORDrawMoveOrCopyObject (hidGC gc)
                 min_multiplier = MIN (min_multiplier, multiplier);
                 max_multiplier = MAX (max_multiplier, multiplier);
               }
-
-            ptr++;
-            i--;
           }
 
 
@@ -597,16 +595,15 @@ XORDrawMoveOrCopyObject (hidGC gc)
   i = Crosshair.AttachedObject.RubberbandN;
   ptr = Crosshair.AttachedObject.Rubberband;
 //  while (i)
-  while (i)
+  for (; i; ptr++, i--)
     {
       PointType *point1, *point2;
 
+      /* this is a rat going to a polygon.  do not draw for rubberband */;
       if (TEST_FLAG (VIAFLAG, ptr->Line))
-	{
-	  /* this is a rat going to a polygon.  do not draw for rubberband */;
-	}
-      else if (TEST_FLAG (RUBBERENDFLAG, ptr->Line))
-//      else
+        continue;
+
+      if (TEST_FLAG (RUBBERENDFLAG, ptr->Line))
 	{
           if (Crosshair.AttachedObject.Type == LINE_TYPE)
             {
@@ -658,9 +655,6 @@ XORDrawMoveOrCopyObject (hidGC gc)
                              ptr->Line->Point2.X + dx,
                              ptr->Line->Point2.Y + dy, ptr->Line->Thickness);
 #endif
-
-      ptr++;
-      i--;
     }
 }
 
diff --git a/src/move.c b/src/move.c
index 6282e9e..cfdf881 100644
--- a/src/move.c
+++ b/src/move.c
@@ -874,7 +874,8 @@ MoveObjectAndRubberband (int Type, void *Ptr1, void *Ptr2, void *Ptr3,
   ptr = Crosshair.AttachedObject.Rubberband;
   while (Crosshair.AttachedObject.RubberbandN)
     {
-      if (Type == LINE_TYPE)
+      if (Type == LINE_TYPE &&
+          ptr->Layer != NULL)
         {
           double x, y, multiplier;
